### PR TITLE
MPA-5596: add KPI-s to gateway: connections meter, req/resp meters, r…

### PR DESCRIPTION
**Pull request by snripa**
…oundtrip time

### CODE REVIEW

**Comment by snripa** [comment]
Guys, if anyone can advice why client discnnection callback doesn't work I'll be happy... @artem-v @aharonha @segabriel @ronenhamias

File: RSocketWebsocket/src/main/java/io/scalecube/gateway/rsocket/websocket/RSocketWebsocketAcceptor.java [hunk 1]
```diff
@@ -67,9 +109,12 @@ private Payload toPayload(ServiceMessage serviceMessage) {
         return codec.encodeAndTransform(serviceMessage, ByteBufPayload::create);
       }
 
+
+
       @Override
       public Mono<Void> onClose() {
         LOGGER.error("------------------> onClose");
```
Comment (artem-v) [inline RSocketWebsocket/src/main/java/io/scalecube/gateway/rsocket/websocket/RSocketWebsocketAcceptor.java]: Please change this `.error()`  to `.info()`

File: RSocketWebsocketRunner/src/main/java/io/scalecube/gateway/rsocket/websocket/RSocketWebsocketGatewayRunner.java [hunk 2]
```diff
@@ -5,22 +5,43 @@
...
+import java.io.File;
 import java.util.Collections;
 import java.util.List;
+import java.util.concurrent.TimeUnit;
 
 public class RSocketWebsocketGatewayRunner {
 
   private static final String SEEDS = "SEEDS";
   private static final List<String> DEFAULT_SEEDS = Collections.singletonList("localhost:4802");
+  public static final String REPORTER_PATH = "metrics";
```
Comment (artem-v) [inline RSocketWebsocketRunner/src/main/java/io/scalecube/gateway/rsocket/websocket/RSocketWebsocketGatewayRunner.java]: Please change to `reports/gw/metrics`

File: RSocketWebsocketRunner/src/main/java/io/scalecube/gateway/rsocket/websocket/RSocketWebsocketGatewayRunner.java [hunk 3]
```diff
@@ -5,22 +5,43 @@
...
+    if (!reporterDir.exists()) {
+      reporterDir.mkdir();
+    }
+
+    CsvReporter csvReporter = CsvReporter.forRegistry(metrics)
+        .convertDurationsTo(TimeUnit.MILLISECONDS)
+        .convertRatesTo(TimeUnit.SECONDS)
+        .build(reporterDir);
+
+    csvReporter.start(10, TimeUnit.SECONDS);
```
Comment (artem-v) [inline RSocketWebsocketRunner/src/main/java/io/scalecube/gateway/rsocket/websocket/RSocketWebsocketGatewayRunner.java]: Cool but where it stops? Maybe put stopping call at shutdown hook ? WDYT ?

File: RSocketWebsocket/src/main/java/io/scalecube/gateway/rsocket/websocket/RSocketWebsocketAcceptor.java [hunk 4]
```diff
@@ -23,40 +27,78 @@
...
+    disconnectedMeter = metrics.getMeter("rsocket", "client", "disconnected");
+    clientReqMeter = metrics.getMeter("rsocket", "client", "requests");
+    clientResponseMeter = metrics.getMeter("rsocket", "client", "responses");
+
   }
 
   @Override
   public Mono<RSocket> accept(ConnectionSetupPayload setup, RSocket rSocket) {
     LOGGER.info("Accepted rSocket websocket: {}, connectionSetup: {}", rSocket, setup);
+    rSocket.onClose().subscribe(s -> disconnectedMeter.mark());
```
Comment (artem-v) [inline RSocketWebsocket/src/main/java/io/scalecube/gateway/rsocket/websocket/RSocketWebsocketAcceptor.java]: Don't you think that we update this meter twice ?   Looks like at [this line](https://github.com/scalecube/scalecube-gateway/pull/87/files#diff-993a8da92c4308a275ce4cc2b6ebdb49R117) we already doing decrement.

File: RSocketWebsocket/src/main/java/io/scalecube/gateway/rsocket/websocket/RSocketWebsocketAcceptor.java [hunk 5]
```diff
@@ -23,40 +27,78 @@
...
 
-  public RSocketWebsocketAcceptor(ServiceCall serviceCall) {
+  public RSocketWebsocketAcceptor(ServiceCall serviceCall, Metrics metrics) {
     this.serviceCall = serviceCall;
+    callRoundtripTime = metrics.getTimer("rsocket", "roundtrip");
+    connectedMeter = metrics.getMeter("rsocket", "client", "connected");
+    disconnectedMeter = metrics.getMeter("rsocket", "client", "disconnected");
+    clientReqMeter = metrics.getMeter("rsocket", "client", "requests");
+    clientResponseMeter = metrics.getMeter("rsocket", "client", "responses");
+
```
Comment (artem-v) [inline RSocketWebsocket/src/main/java/io/scalecube/gateway/rsocket/websocket/RSocketWebsocketAcceptor.java]: `connectedMeter` and `disconnectedMeter` :  1) why they are meters and not counters?  2) why not have  just one counter called activeConnections?

File: RSocketWebsocketRunner/src/main/java/io/scalecube/gateway/rsocket/websocket/RSocketWebsocketGatewayRunner.java (hunk 3 above)
Comment (segabriel) [reply RSocketWebsocketRunner/src/main/java/io/scalecube/gateway/rsocket/websocket/RSocketWebsocketGatewayRunner.java]: yeah, and add invoking `csvReporter.report()` in a shutdown callback, to don't lose latest metrics

File: WebsocketRunner/src/main/java/io/scalecube/gateway/websocket/WebsocketGatewayRunner.java [hunk 6]
```diff
@@ -36,15 +41,34 @@ public static void main(String[] args) throws InterruptedException {
...
+    if (!reporterDir.exists()) {
+      //noinspection ResultOfMethodCallIgnored
+      reporterDir.mkdirs();
+    }
+    CsvReporter csvReporter = CsvReporter.forRegistry(metrics)
+      .convertDurationsTo(TimeUnit.MILLISECONDS)
+      .convertRatesTo(TimeUnit.SECONDS)
+      .build(reporterDir);
+
+    csvReporter.start(10, TimeUnit.SECONDS);
```
Comment (segabriel) [inline WebsocketRunner/src/main/java/io/scalecube/gateway/websocket/WebsocketGatewayRunner.java]: add after this the shutdown hook and invoke `csvReporter.report()` in it.

**Comment by artem-v** [comment]
> Guys, if anyone can advice why client discnnection callback doesn't work I'll be happy... @artem-v @aharonha @segabriel @ronenhamias

Question was cleared.

**Pull request merged**